File: exercise/week04/php/src/Construct/ArtefactUpdaterFactory.php
<?php

namespace Construct;

class ArtefactUpdaterFactory
{
    private const AGED_SIGNAL = "Aged Signal";
    private const BACKDOOR_PASS = "Backdoor Pass to TAFKAL80ETC Protocol";
    private const SULFURAS_CORE = "Sulfuras Core Fragment";
    private const STEALTH_CLOAK = "Stealth Cloak v2.0";

    public static function forArtefact(string $name): ArtefactUpdater
    {
        return match ($name) {
            self::AGED_SIGNAL => new AgedSignalUpdater(),
            self::BACKDOOR_PASS => new BackdoorPassUpdater(),
            self::SULFURAS_CORE => new SulfurasUpdater(),
            self::STEALTH_CLOAK => new StealthCloakUpdater(),
            default => new NormalArtefactUpdater(),
        };
    }
}

File: exercise/week04/java/src/main/java/construct/ArtefactUpdaterFactory.java
package construct;

public class ArtefactUpdaterFactory {
    private static final String AGED_SIGNAL = "Aged Signal";
    private static final String BACKDOOR_PASS = "Backdoor Pass to TAFKAL80ETC Protocol";
    private static final String SULFURAS_CORE = "Sulfuras Core Fragment";
    private static final String STEALTH_CLOAK = "Stealth Cloak v2.0";

    private ArtefactUpdaterFactory() {
    }

    public static ArtefactUpdater forArtefact(String name) {
        return switch (name) {
            case AGED_SIGNAL -> new AgedSignalUpdater();
            case BACKDOOR_PASS -> new BackdoorPassUpdater();
            case SULFURAS_CORE -> new SulfurasUpdater();
            case STEALTH_CLOAK -> new StealthCloakUpdater();
            default -> new NormalArtefactUpdater();
        };
    }
}